src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,267 @@
+import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.world.entity.EquipmentSlot;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.flag.FeatureFlagSet;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.enchantment.EnchantmentEffectComponents;
+import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import net.minecraft.world.level.ItemLike;
+import net.neoforged.neoforge.capabilities.ItemCapability;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Immutable combination of an {@link Item} and data components.
+ * Similar to an {@link ItemStack}, but immutable and without a count.
+ */
+public final class ItemResource implements IDataComponentHolderResource<Item> {
+    /**
+     * Codec for an item resource.
+     * Same format as {@link ItemStack#SINGLE_ITEM_CODEC}.
+     * Does <b>not</b> accept empty resources.
+     */
+    public static final Codec<ItemResource> CODEC = Codec.lazyInitialized(() -> ItemStack.SINGLE_ITEM_CODEC.xmap(ItemResource::of, ItemResource::toStack));
+    /**
+     * Codec for an item resource. Same format as {@link #CODEC}, and also accepts empty resources.
+     */
+    public static final Codec<ItemResource> OPTIONAL_CODEC = ExtraCodecs.optionalEmptyMap(CODEC).xmap(ItemResource::fromOptional, ItemResource::asOptional);
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private static ItemResource fromOptional(Optional<ItemResource> optional) {
+        return optional.orElse(ItemResource.EMPTY);
+    }
+
+    private Optional<ItemResource> asOptional() {
+        return isEmpty() ? Optional.empty() : Optional.of(this);
+    }
+
+    /**
+     * Stream codec for an item resource. Accepts empty resources.
+     */
+    public static final StreamCodec<RegistryFriendlyByteBuf, ItemResource> STREAM_CODEC = StreamCodec.composite(
+            ByteBufCodecs.holderRegistry(Registries.ITEM), ItemResource::getHolder,
+            DataComponentPatch.STREAM_CODEC, ItemResource::getComponentsPatch,
+            ItemResource::of);
+
+    public static ItemStack itemStackOf(IResourceStack<ItemResource> resourceStack) {
+        return resourceStack.resource().toStack(resourceStack.amount());
+    }
+
+    public static final ItemResource EMPTY = new ItemResource(ItemStack.EMPTY);
+    public static final ResourceStack<ItemResource> EMPTY_STACK = new ResourceStack<>(EMPTY, 0);
+
+    /**
+     * This is used only for registry, you should not use this method!
+     */
+    @ApiStatus.Internal
+    public static ItemResource invalidateDefault(ItemLike item) {
+        return item == Items.AIR ? EMPTY : new ItemResource(item.asItem().getDefaultInstance().copyWithCount(1));
```

Comment:
the `item == Items.AIR` check here should be invoking `asItem()`
This would fix potential issue with passing `Blocks.AIR` not returning `EMPTY` since `Blocks.AIR` is not `Items.AIR` while `Blocks.AIR.asItem()` does result in `Items.AIR`

If `Block` (and other none `ItemLike` objects) are not supposed to be passed here maybe worth changing to `Item` instead of `ItemLike`?

Do note it is common for custom holder types to be `ItemLike` instances, allowing these holders to passed aound as if they were the backing the `Item` type. (See `DeferredBlock`/`DeferredItem`)

```suggestion
        return item.asItem() == Items.AIR ? EMPTY : new ItemResource(item.asItem().getDefaultInstance().copyWithCount(1));
```